### PR TITLE
Integration of Lightning 

### PR DESCRIPTION
**Pull request by RBirmiwal**

### CODE REVIEW

**Comment by RBirmiwal** [comment]
@madelynshapiro 

## Ready For Review 
**Example Notebooks**
- [ ] other_examples/lightning_custom_training_example.ipynb
- [ ] other_examples/lightning_cvxpy_layers.ipynb
- [ ] Part_1_lightning_basics_tutorial.ipynb
- [ ] Part_2_lightning_advanced...
- [ ] Part_3_multi_GPU_example.py
- [ ] README.md

** Source Code**
- [ ] constraint.py
- [ ] problem.py
- [ ] trainer.py
- [ ] dataset.py
- [ ] utils.py
- [ ] __init__.py
- [ ] 

## To Be Done
** For notebooks and README**
- [ ]  Add separate logging example (Tensorboard and WandB)
- [ ]  Add wandb hyperparameter tuning example (learning rate, batch_size, and sizes within blocks.MLP). Will use NSSM_building_dynamics.ipynb as demo bed 
- [ ] Add wandb information to README
- [ ] Reformat/edit DeepKoopman notebook (will use this as demo bed for tensorboard, as is currently) and PINN_DiffusionEquation notebook
** For source code **
- [ ] Fix warmup in trainer.py
- [ ] Update as required to trainer/problem/dataset.py (e.g. such that hyperparameters can be driven by config)
- [ ] Update pyproject.toml to include PyTorch Lightning install

**Comment by RBirmiwal** [comment]
@madelynshapiro @drgona This is ready for review. Based off feedback from the demo, I will include how to optimize Neuromancer component hyperparameters using wandb (e.g. blocks.MLP sizes)/other things in next release. For now I have included wandb hyperparameter tuning for only batch size and learning rate. To do the latter Neuromancer problem setups need to be based off config or to wrap problem formulation within a "problem_setup_function" function that is driven by a config.

File: pyproject.toml [hunk 1]
```diff
@@ -48,7 +48,8 @@ dependencies = [
     "pyts",
     "torch",
     "torchdiffeq",
-    "toml"
+    "toml", 
+    "lightning"
```
Comment (madelynshapiro) [inline pyproject.toml:52]: Let's make lightning an optional dependency under `[project.optional-dependencies]` as `lightning = ["lightning"]`

File: pyproject.toml [hunk 2]
```diff
@@ -48,7 +48,8 @@ dependencies = [
...
 ]
 
 version = "1.4.2"
```
Comment (madelynshapiro) [inline pyproject.toml:55]: Do we need to update this manually or will it be bumped automatically on release?

[the remark this answers is not in the document]
Comment (madelynshapiro) [reply pyproject.toml:55]: nevermind, we can handle this in the PR from dev->main

File: pyproject.toml (hunk 1 above)
Comment (drgona) [reply pyproject.toml:52]: we might need lightning as core dependency because it is imported in constraints.py

**Review by drgona — APPROVED** [review]
great job Rahul!
